// @bochilteam/scraper/lib/texts/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.truthjson = exports.truth = exports.darejson = exports.dare = exports.bucinjson = exports.bucin = void 0;
const bucin_1 = __importStar(require("./bucin"));
exports.bucin = bucin_1.default;
Object.defineProperty(exports, "bucinjson", { enumerable: true, get: function () { return bucin_1.bucinjson; } });
const dare_1 = __importStar(require("./dare"));
exports.dare = dare_1.default;
Object.defineProperty(exports, "darejson", { enumerable: true, get: function () { return dare_1.darejson; } });
const truth_1 = __importStar(require("./truth"));
exports.truth = truth_1.default;
Object.defineProperty(exports, "truthjson", { enumerable: true, get: function () { return truth_1.truthjson; } });
__exportStar(require("./aksarajawa"), exports);
