// @bochilteam/scraper/lib/games/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tebaktebakanjson = exports.tebaktebakan = exports.tebaklirikjson = exports.tebaklirik = exports.tebakbenderajson = exports.tebakbendera = exports.susunkatajson = exports.susunkata = exports.siapakahakujson = exports.siapakahaku = exports.tebakkabupatenjson = exports.tebakkabupaten = exports.tebakkimiajson = exports.tebakkimia = exports.tekatekijson = exports.tekateki = exports.tebakkatajson = exports.tebakkata = exports.asahotakjson = exports.asahotak = exports.family100json = exports.family100 = exports.caklontongjson = exports.caklontong = exports.tebakgambarjson = exports.tebakgambar = void 0;
const asahotak_1 = __importStar(require("./asahotak"));
exports.asahotak = asahotak_1.default;
Object.defineProperty(exports, "asahotakjson", { enumerable: true, get: function () { return asahotak_1.asahotakjson; } });
const caklontong_1 = __importStar(require("./caklontong"));
exports.caklontong = caklontong_1.default;
Object.defineProperty(exports, "caklontongjson", { enumerable: true, get: function () { return caklontong_1.caklontongjson; } });
const family100_1 = __importStar(require("./family100"));
exports.family100 = family100_1.default;
Object.defineProperty(exports, "family100json", { enumerable: true, get: function () { return family100_1.family100json; } });
const siapakahaku_1 = __importStar(require("./siapakahaku"));
exports.siapakahaku = siapakahaku_1.default;
Object.defineProperty(exports, "siapakahakujson", { enumerable: true, get: function () { return siapakahaku_1.siapakahakujson; } });
const susunkata_1 = __importStar(require("./susunkata"));
exports.susunkata = susunkata_1.default;
Object.defineProperty(exports, "susunkatajson", { enumerable: true, get: function () { return susunkata_1.susunkatajson; } });
const tebakbendera_1 = __importStar(require("./tebakbendera"));
exports.tebakbendera = tebakbendera_1.default;
Object.defineProperty(exports, "tebakbenderajson", { enumerable: true, get: function () { return tebakbendera_1.tebakbenderajson; } });
const tebakgambar_1 = __importStar(require("./tebakgambar"));
exports.tebakgambar = tebakgambar_1.default;
Object.defineProperty(exports, "tebakgambarjson", { enumerable: true, get: function () { return tebakgambar_1.tebakgambarjson; } });
const tebakkabupaten_1 = __importStar(require("./tebakkabupaten"));
exports.tebakkabupaten = tebakkabupaten_1.default;
Object.defineProperty(exports, "tebakkabupatenjson", { enumerable: true, get: function () { return tebakkabupaten_1.tebakkabupatenjson; } });
const tebakkata_1 = __importStar(require("./tebakkata"));
exports.tebakkata = tebakkata_1.default;
Object.defineProperty(exports, "tebakkatajson", { enumerable: true, get: function () { return tebakkata_1.tebakkatajson; } });
const tebakkimia_1 = __importStar(require("./tebakkimia"));
exports.tebakkimia = tebakkimia_1.default;
Object.defineProperty(exports, "tebakkimiajson", { enumerable: true, get: function () { return tebakkimia_1.tebakkimiajson; } });
const tebaklirik_1 = __importStar(require("./tebaklirik"));
exports.tebaklirik = tebaklirik_1.default;
Object.defineProperty(exports, "tebaklirikjson", { enumerable: true, get: function () { return tebaklirik_1.tebaklirikjson; } });
const tebaktebakan_1 = __importStar(require("./tebaktebakan"));
exports.tebaktebakan = tebaktebakan_1.default;
Object.defineProperty(exports, "tebaktebakanjson", { enumerable: true, get: function () { return tebaktebakan_1.tebaktebakanjson; } });
const tekateki_1 = __importStar(require("./tekateki"));
exports.tekateki = tekateki_1.default;
Object.defineProperty(exports, "tekatekijson", { enumerable: true, get: function () { return tekateki_1.tekatekijson; } });
